File: android/src/main/java/com/turboimage/TurboImageCacheKeyIndex.kt
package com.turboimage

import android.content.Context

/**
 * Tracks cache identity keys (cacheKey ?: uri) ever used by TurboImage, since
 * Coil's DiskCache exposes no cheap enumerate-by-prefix API - prefix-based
 * clearing needs its own index to know which keys exist.
 */
object TurboImageCacheKeyIndex {
  private const val PREFS_NAME = "TurboImageCacheKeyIndex"
  private const val KEY_SET = "keys"

  @Synchronized
  fun register(context: Context, key: String) {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    val keys = HashSet(prefs.getStringSet(KEY_SET, emptySet()) ?: emptySet())
    if (keys.add(key)) {
      prefs.edit().putStringSet(KEY_SET, keys).apply()
    }
  }

  @Synchronized
  fun keys(context: Context, includePrefix: String?, excludePrefix: String?): List<String> {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    val keys = prefs.getStringSet(KEY_SET, emptySet()) ?: emptySet()
    return keys.filter { key ->
      val included = includePrefix?.let { key.startsWith(it) } ?: true
      val excluded = excludePrefix?.let { key.startsWith(it) } ?: false
      included && !excluded
    }
  }

  @Synchronized
  fun remove(context: Context, keysToRemove: Collection<String>) {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    val keys = HashSet(prefs.getStringSet(KEY_SET, emptySet()) ?: emptySet())
    keys.removeAll(keysToRemove.toSet())
    prefs.edit().putStringSet(KEY_SET, keys).apply()
  }

  @Synchronized
  fun clear(context: Context) {
    context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).edit().clear().apply()
  }
}

File: ios/TurboImageCacheKeyIndex.swift
import Foundation

/// Tracks cache identity keys (cacheKey ?? uri) ever used by TurboImage, since
/// Nuke's DataCache/ImageCache expose no enumerate API — prefix-based clearing
/// needs its own index to know which keys exist.
final class TurboImageCacheKeyIndex {
  static let shared = TurboImageCacheKeyIndex()

  private let defaultsKey = "TurboImageCacheKeyIndex"
  private let queue = DispatchQueue(label: "com.turboimage.cacheKeyIndex")
  private let defaults = UserDefaults.standard

  private init() {}

  func register(_ key: String) {
    queue.sync {
      var keys = storedKeys()
      if keys.insert(key).inserted {
        defaults.set(Array(keys), forKey: defaultsKey)
      }
    }
  }

  func keys(includePrefix: String?, excludePrefix: String?) -> [String] {
    queue.sync {
      storedKeys().filter { key in
        let included = includePrefix.map { key.hasPrefix($0) } ?? true
        let excluded = excludePrefix.map { key.hasPrefix($0) } ?? false
        return included && !excluded
      }
    }
  }

  func remove(_ keys: [String]) {
    queue.sync {
      var stored = storedKeys()
      keys.forEach { stored.remove($0) }
      defaults.set(Array(stored), forKey: defaultsKey)
    }
  }

  func removeAll() {
    queue.sync {
      defaults.removeObject(forKey: defaultsKey)
    }
  }

  private func storedKeys() -> Set<String> {
    Set(defaults.stringArray(forKey: defaultsKey) ?? [])
  }
}
